Improve errors for TOML fields that support workspace inheritance

rdimartino opened the pull request:
Fixes #10997

This also addresses the issue with `MaybeWorkspace<VecStringOrBool>` mentioned in #10942:

```
Caused by:
  invalid type: string "foo", expected a boolean or vector of strings for key `package.publish`
```

I removed the `maybe_workspace_vec_string` deserializer in 7a50c0c718f3547e3d76498770fc3ed256cb3994 because the error message from the inner `Vec<String>` is now surfaced, but I can revert that if it's preferable to keep those changes.

I tried to base the `deserialize` implementation off of the derived impl for an untagged enum from `cargo expand`. This approach [ultimately led me](https://github.com/serde-rs/serde/blob/v1.0.144/serde/src/private/de.rs#L218) to adding the `serde-value` dependency.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @epage (or someone else) soon.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

weihanglo reviewed (commented):
Thank you for this pull request.

It seems that the test case `cargo_metadata_with_invalid_version_field` is also valid even without this patch. Does that mean the change is not relevant anymore? If in this case, I suggest leave the test here in PR to prevent regression and remove other parts.

rdimartino commented:
@weihanglo Ah, you're right! Sorry about that. I overlooked that `version` has `deserialize_with = "version_trim_whitespace"` so that test that I added didn't actually test my patch. I've added a test for the field `publish: Option<MaybeWorkspace<VecStringOrBool>>` which shows the improvement.

Before:
```
$ cargo metadata --verbose --format-version 1 --all-features --manifest-path /tmp/test/Cargo.toml
error: failed to parse manifest at `/private/tmp/test/Cargo.toml`

Caused by:
  data did not match any variant of untagged enum MaybeWorkspace for key `package.publish`
```

After:
```
$ ./target/debug/cargo metadata --verbose --format-version 1 --all-features --manifest-path /tmp/test/Cargo.toml
error: failed to parse manifest at `/tmp/test/Cargo.toml`

Caused by:
  invalid type: string "foo", expected a boolean or vector of strings for key `package.publish`
```

Muscraft commented on Cargo.toml:
```diff
@@ -52,6 +52,7 @@ semver = { version = "1.0.3", features = ["serde"] }
 serde = { version = "1.0.123", features = ["derive"] }
 serde_ignored = "0.1.0"
 serde_json = { version = "1.0.30", features = ["raw_value"] }
+serde-value = "0.7.0"
```
I am worried about adding this as a dependency since it hasn't been updated since [July 15th, 2020](https://github.com/arcnmx/serde-value) and the comment you linked to about `serde-value` appears to be [6 years old](https://github.com/serde-rs/serde/blame/f52d134c14f6904010d0a59107987b050b36b811/serde/src/private/de.rs#L218)

Someone on the [Cargo team](https://www.rust-lang.org/governance/teams/dev-tools#cargo) could have a different opinion so I would wait for them to comment as well.

epage replied:
`serde-value` seems innocuous enough and we can always vendor it if needed.  I think how much it cleans up this code shows it pulls its weight.

epage commented:
@bors r+

Looks like @weihanglo 's test concern  was resolved so this seems good to go.  Thanks for doing this!

bors commented:
:pushpin: Commit 7b16e59896fc44ce23d24ff8dde2e83382e1be5b has been approved by `epage`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=epage 7b16e59896fc44ce23d24ff8dde2e83382e1be5b -->
<!-- homu: {"type":"Approved","sha":"7b16e59896fc44ce23d24ff8dde2e83382e1be5b","approver":"epage","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 7b16e59896fc44ce23d24ff8dde2e83382e1be5b with merge 9230e4839dad2f71080ed9151f4400cc8de362b9...
<!-- homu: {"type":"BuildStarted","head_sha":"7b16e59896fc44ce23d24ff8dde2e83382e1be5b","merge_sha":"9230e4839dad2f71080ed9151f4400cc8de362b9"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3108426921/jobs/5038090686)
Approved by: epage
Pushing 9230e4839dad2f71080ed9151f4400cc8de362b9 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"epage","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3108426921/jobs/5038090686"},"merge_sha":"9230e4839dad2f71080ed9151f4400cc8de362b9"} -->
